hpx/runtime/applier/apply_helper.hpp

```diff
@@ -134,7 +135,10 @@ namespace hpx { namespace applier { namespace detail
             naming::address::address_type lva,
             threads::thread_priority priority, Ts &&... vs)
         {
-            if (this_thread::has_sufficient_stack_space() || hpx::is_pre_startup())
+            // Direct actions should be able to be executed from a non-HPX thread
+            // as well
+            if (this_thread::has_sufficient_stack_space() ||
+                hpx::threads::get_self_ptr() == nullptr)
```

Comment:
Here you check both, the stack space and whether its running on an HPX thread. You might want to revisit my comment above related to this.